File: hexameter-core/src/test/java/org/codetome/hexameter/core/StressTest.java
package org.codetome.hexameter.core;

import org.codetome.hexameter.core.api.Hexagon;
import org.codetome.hexameter.core.api.HexagonalGrid;
import org.codetome.hexameter.core.api.HexagonalGridBuilder;
import org.junit.Before;
import org.junit.Test;
import rx.functions.Action1;

import java.util.concurrent.atomic.AtomicInteger;

import static java.lang.String.format;
import static org.assertj.core.api.Assertions.assertThat;

/**
 * This should be later refactored to a stress test directory.
 */
public class StressTest {

    private static final int BIG_GRID_HEIGHT = 1000;
    private static final int BIG_GRID_WIDTH = 1000;
    private static final double BIG_GRID_RADIUS = 50;
    private static final long EXPECTED_MAXIMUM_GENERATION_TIME = 2000;
    private static final long EXPECTED_MAXIMUM_FETCH_TIME = 2000;

    private HexagonalGridBuilder hexagonalGridBuilder;

    @Before
    public void setUp() {
        hexagonalGridBuilder = new HexagonalGridBuilder()
                .setGridHeight(BIG_GRID_HEIGHT)
                .setGridWidth(BIG_GRID_WIDTH)
                .setRadius(BIG_GRID_RADIUS);
    }

    @Test
    public void shouldBeAbleToCreateBigGrid() {
        final long start = System.nanoTime();
        hexagonalGridBuilder.build();
        final long end = System.nanoTime();
        assertThat(nanoToMs(end - start)).isLessThan(EXPECTED_MAXIMUM_GENERATION_TIME);
    }

    @Test
    public void shouldBeAbleToFetchHexesFromBigGrids() {
        HexagonalGrid grid = hexagonalGridBuilder.build();

        final long start = System.nanoTime();
        final AtomicInteger ai = new AtomicInteger();
        grid.getHexagons().forEach(new Action1<Hexagon>() {
            @Override
            public void call(Hexagon hexagon) {
                ai.incrementAndGet();
            }
        });
        final long end = System.nanoTime();
        System.out.println(format("Number of hexes: %s, generated in: %dms.", ai.get(), nanoToMs(end - start)));
        assertThat(nanoToMs(end - start)).isLessThan(EXPECTED_MAXIMUM_FETCH_TIME);
    }

    private long nanoToMs(long nanoTime) {
        return nanoTime / 1000 / 1000;
    }
}

File: mixite.core/jvm/src/test/kotlin/org/hexworks/mixite/core/StressTest.kt
package org.hexworks.mixite.core

import org.hexworks.mixite.core.api.HexagonalGridBuilder
import org.hexworks.mixite.core.api.contract.SatelliteData
import org.junit.Test
import java.lang.String.format
import java.util.concurrent.atomic.AtomicInteger
import kotlin.test.BeforeTest
import kotlin.test.assertTrue

class StressTest {

    lateinit var hexagonalGridBuilder: HexagonalGridBuilder<out SatelliteData>

    @BeforeTest
    fun setUp() {
        hexagonalGridBuilder = HexagonalGridBuilder<SatelliteData>()
                .setGridHeight(BIG_GRID_HEIGHT)
                .setGridWidth(BIG_GRID_WIDTH)
                .setRadius(BIG_GRID_RADIUS)
    }

    @Test
    fun shouldBeAbleToCreateBigGrid() {
        val start = System.nanoTime()
        hexagonalGridBuilder.build()
        val end = System.nanoTime()
        assertTrue {
            nanoToMs(end - start) < EXPECTED_MAXIMUM_GENERATION_TIME
        }
    }

    @Test
    fun shouldBeAbleToFetchHexesFromBigGrids() {
        val grid = hexagonalGridBuilder.build()

        val start = System.nanoTime()
        val ai = AtomicInteger()
        for (o in grid.hexagons) {
            ai.incrementAndGet()
        }
        val end = System.nanoTime()
        println(format("Number of hexes: %s, generated in: %dms.", ai.get(), nanoToMs(end - start)))
        assertTrue {
            nanoToMs(end - start) < EXPECTED_MAXIMUM_FETCH_TIME
        }
    }

    private fun nanoToMs(nanoTime: Long): Long {
        return nanoTime / 1000 / 1000
    }

    companion object {

        private const val BIG_GRID_HEIGHT = 1000
        private const val BIG_GRID_WIDTH = 1000
        private const val BIG_GRID_RADIUS = 50.0
        private const val EXPECTED_MAXIMUM_GENERATION_TIME: Long = 3000
        private const val EXPECTED_MAXIMUM_FETCH_TIME: Long = 3000
    }
}
